### DISCUSSION TITLE
Implementing NOMA, OTFS into Sionna

### DISCUSSION THREAD

**Discussion by cacent96** [NVlabs/sionna#56] [Q&A] [answered]
Hello Guys,

I'm trying to create NOMA and OTFS model and use it in Sionna, Any comment or suggestion how I can do this?

**Comment by jhoydis** [comment, marked as answer]
Hi,

Sionna does not yet provide an OTFS module that you can use. 

Implementing it should not be difficult and would require roughly the following steps:
- Define an OTFSResourceGrid which determines how pilots and data symbols are distributed in time and frequency. This would probably look similar to the ``sionna.ofdm.ResourceGrid``. Maybe it could even be used as is.
- Implement an OTFSModulator that maps a batch of constellation symbols to time-domain IQ samples that can be sent over a TimeChannel (again probably similar to the ``sionna.ofdm.OFDMModulator``). This could be very simple in case that you want to implement OTFS on top of OFDM (probably you would only need to compute an FFT over the OFDM resource grid). However, as far as I know, there are multiple ways of how OTFS can be implemented.
- Implement an OTFSDemodulator that transforms a vector of received time-domain samples back to the OTFSResourceGrid and implement methods for channel estimation and symbol detection for OTFS. This is probably where it gets a bit complicated. It all depends on which algorithms you want to implement.

Hope that helps to get you started.

**Reply by vikkicoder** [reply]
Any suggestions on creating delay Doppler channel model similar to cir_to_time or cir_to_freq functions?
